### ISSUE TITLE
Erroneous assignments on 10k PBMC from 10x 

### ISSUE THREAD

**Issue by apredeus** [deweylab/CellO#12]
Hi, 

Thank you for developing this tool - I think this is a very neat and useful idea, especially given how Cell Ontology seems to be gaining traction. I included it in the `Scanpy` tutorial I'm putting together at our Institute that involves various things that are not covered by the vignettes on the `Scanpy` website. I was using 10k PBMC dataset I obtained from [10x Genomics](https://cf.10xgenomics.com/samples/cell-exp/4.0.0/Parent_NGSC3_DI_PBMC/Parent_NGSC3_DI_PBMC_filtered_feature_bc_matrix.h5) 

I noticed that a couple of fairly well-defined cell types are annotated with errors. Here's the "most specific cell type" annotation plot: 

![image](https://user-images.githubusercontent.com/7825825/142686485-8947adb8-33b5-4ddc-a1cb-ec7c5ab0ec82.png)

These, however, are the classical markers used for the PBMC sub-populations. FCGR3A is non-classical monocytes, FCER1A is myeloid DCs, and LILRA4 is plasmacytoid DCs:

![image](https://user-images.githubusercontent.com/7825825/142687247-e246ea2e-82af-4b62-ad40-a45ad38dc6ac.png)

You can see that the model mistook non-classical monocytes for dendritic cells, and annotated both subtypes of DCs as "mononuclear cell" which is probably not what's intended too.

I can provide the whole notebook if that would make it easier to diagnose the problem (but I did pretty much follow the tutorial, with the addition of `.todense()` transformation).

**Comment by mbernste** [comment]
Hi, thanks for your interest in CellO! Im sorry to hear that it’s making these mistakes. I do have a few thoughts/suggestions which I will outline below:

1. The first thing I’ll note is that CellO is not completely accurate and it’s accuracy varies across cell types. We intend to improve its training set over time, but it is not unusual for it to confuse similar cell types (for example, dendritic cells and monocytes have similar expression profiles). I would think of CellO as a (hopefully) relatively painless method for getting a first draft annotation, especially when you have diverse cell types that might not have canonical markers (such as cells from a full piece of tissue rather than just PBMC’s). I’d definitely double check the results when you do have cell types with canonical markers as you are doing here
2. That said, I usually don’t see CellO incorrectly label dendritic cells as monocytes. Furthermore it usually accurately labels plasmocytoid dendritic cells. I have a few suggestions to try to figure out why it’s making these errors: 1) do you mind showing me the clusters it’s using? I wonder if the clusters are either too small or too big. 2) How many genes are you using? Did you filter for highly variable genes? I suggest using as many genes as possible.
3. The label “mononuclear cell” is actually a cell type that is higher up in the Cell Ontology so it’s labeling of these cells as mononuclear cell is not strictly wrong, but it obviously isn’t too informative 
4. I would be curious to see the output of the function ‘scanpy_cello.cello_probs’ when run on the clusters that it erred on. This function will plot the ontology graph colored by the probability the cluster is of each cell type. For example, I wonder what probabilities it is giving to the dendritic cell cluster.
5. If you don’t mind sharing your analysis, I could double check to see if I spot anything else that might be causing issues

Best,
Matt

**Comment by apredeus** [comment]
Hi Matt, 

Thank you for your clarifications! I understand that `CellO` is not perfectly accurate - actually from my experience none of the existing tools come even close to being that. But I did like many things about your tool though - especially the "native" Cell Ontology support which will become increasingly useful with `cellxgene` and other standardization efforts. 

This is the complete notebook that I've used - you should be able to see it (the `CellO` part is on the very bottom): https://github.com/cellgeni/notebooks/blob/master/notebooks/new-10kPBMC-Scanpy.ipynb

For simplicity, here's the `h5ad` object that exactly matches `new_adata` before any normalization for running `CellO` (after cell #97): https://www.dropbox.com/s/ar06daabnejdnj6/10k_pbmc_filtered.h5ad

Let me know if this is enough to reproduce the analysis. Thank you!
-- Alex

**Comment by mbernste** [comment]
Hi Alex,

I was more or less able to reproduce your analysis; my UMAP plot looks slightly different, but overall I was able to reproduce the clusters and classifications. For reference, here are my UMAP plots with the marker genes of interest:

![image](https://user-images.githubusercontent.com/5004419/143151251-6932c26d-7229-4731-a635-9810cfbf7a3b.png)

Overall, I was at least able to figure out why the true dendritic cells were only being labeled as "mononuclear cells".  Essentially, it comes down to the way in which CellO chooses the most specific cell type: if there are multiple low-level cell types with high probability, then it picks the cell type with _highest_ probability (see the schematic from the paper below):  

![image](https://user-images.githubusercontent.com/5004419/143148817-9f99cff5-68a9-4719-8f0e-82a65c0f5c08.png)

When I look at the probabilities assigned to the true conventional dendritic cells, you can see that CellO is indeed assigning them a high probability, but not as high as 'mononuclear cell', which is located in a different region of the ontology:

![image](https://user-images.githubusercontent.com/5004419/143150344-12e102e8-49dc-4b66-a7b4-a2f274dd66bd.png)

The same goes for the plasmocytoid dendritic cells:

![image](https://user-images.githubusercontent.com/5004419/143150403-dc647dd6-0032-40b1-80af-7ae333acfa31.png)

 Thus, CellO chooses 'mononuclear cell' instead of 'conventional dendritic cell' or 'plasmocytoid dendritic cell' as the most specific terms for these clusters. This error is basically due to the sort of strange way that the Cell Ontology is structured in regards to the term "mononuclear cell". 

For now, there is not a great way to fix this other than looking at the actual probabilities assigned to the cell types (either by coloring the cells according to "plasmacytoid dendritic cell (probability)" or by viewing all of the probabilities assigned to the "mononuclear cell" cluster via the `scanpy_cello.cello_probs ` function:

![image](https://user-images.githubusercontent.com/5004419/143150698-aefb951a-b6b1-4bd0-981d-f612bd1afc31.png)

In the next version of CellO, I may consider hardcoding a rule into CellO that will prefer an alternative term over "mononuclear cell" when choosing a most specific label.

Lastly, in regards to the non-classical monocytes being classified as "dendritic cell", this appears to me to just be an unfortunate mistake by the classifier (while . I do know that CellO's training set does not have many non-classical monocyte samples and I believe that the lack of CD14 expression coupled with overall myeloid expression signature leads the classifier to think its a cluster of dendritic cells; though you can see it is less confident about that call:

![image](https://user-images.githubusercontent.com/5004419/143151039-a11c7284-06c4-4537-af60-538ec7a52df2.png)

I will have to hunt down some more non-classical monocyte samples to add the training set for the next iteration.

Please let me know if I can clarify any of this or if you have any more questions.

Best,
Matt

**Comment by apredeus** [comment]
Hi Matt, 

Thank you very much for looking into this. Your explanations make perfect sense. I'll check for the updated reference later too. 

Cheers

-- Alex
